Apex7000_BillValidator/Serial/SlaveDataValidator.cs

```diff
@@ -0,0 +1,224 @@
+        /// property with every error that is parsed. Adds a validation message fragment
+        /// indicating what errors were parsed from the raw data.
+        /// </summary>
+        protected void CheckErrors()
+        {
+            if ((Data[2] & 0b00000001) != 0)
+            {
+                Errors.Add("power up");
+            }
+            if ((Data[2] & 0b00000010) != 0)
+            {
+                Errors.Add("invalid command");
+            }
+            if ((Data[2] & 0b00000100) != 0)
+            {
+                Errors.Add("ba failed");
+            }
+
+            if (Errors.Any())
+            {
+                var errorString = $"Errors: {string.Join(",", Errors.ToArray())}";
+                ValidationMessageFragments.Add(errorString);
+            }
+        }
+        
+        /// <summary>
+        /// Bill value that has been credited, if any
+        /// </summary>
+        protected string BillValue = string.Empty;
+
+        /// <inheritdoc/>
+        protected override bool ValidateSerialData()
+        {
+            ValidationMessageFragments.Add("SLAVE");
+            if (RawData == null)
+            {
+                ValidationMessageFragments.Add("Raw data is null");
+                return false;
+            }
+
+            if (RawData.Length == 0)
+            {
+                ValidationMessageFragments.Add("No response");
+                return false;
+            }
+
+            // add the raw bytes being parsed to our message
+            {
+                var rawBytes = string.Empty;
+                foreach (var b in RawData)
+                {
+                    rawBytes += "0x" + Convert.ToString(b, 16) + " ";
+                }
+            
+                ValidationMessageFragments.Add($"Raw Bytes received: {rawBytes}");
+            }
+
+            if (RawData.Length < 2 || RawData.Length != RawData[1])
+            {
+                ValidationMessageFragments.Add("Message received too short");
+                return false;
+            }
+            
+            if ((RawData[2] & 0x20) != 0x20)
+            {
+                ValidationMessageFragments.Add("Slave did not set slave bit");
+                return false;
+            }
+
+            ValidationMessageFragments.Add($"ACK {RawData[2] & 0b00001111}");
+            Data = RawData.Skip(3).Take(6).ToArray();
+            
+
+            if (!CheckChecksum())
+            {
+                ValidationMessageFragments.Add("Bad checksum");
+                return false;
+            }
+
+            CheckStates();
+
+            if (IsNoState && ((Data[1] & 0b1110111) == 0))
+            {
+                ValidationMessageFragments.Add("Both state (byte 0) and error (byte 1) bytes are zero. " +
+                                               "The state byte should only equal zero if there are error bits set.");
+                return false;
```

Comment:
I don't think we should return yet. Maybe just a running flag you modify. `okay = okay && SomeCheck()` so that once okay is set to false, it cannot be set to true again.